packages/gatsby/src/utils/node-manifest.ts

```diff
@@ -0,0 +1,303 @@
+
+  // but if we have more than one page where this node shows up
+  // we need to try to be more specific
+  if (pagePathSetOrMap && pagePathSetOrMap.size > 1) {
+    let ownerPagePath: string | undefined
+    let foundOwnerNodeId = false
+
+    // for each page this nodeId is queried in
+    for (const pathOrPageObject of pagePathSetOrMap.values()) {
+      // if we haven't found a page with this nodeId
+      // set as page.ownerNodeId then run this logic.
+      // this condition is on foundOwnerNodeId instead of ownerPagePath
+      // in case we find a page with the nodeId in page.context.id
+      // and then later in the loop there's a page with this nodeId
+      // set on page.ownerNodeId.
+      // We always want to prefer ownerPagePath over context.id
+      if (foundOwnerNodeId) {
+        break
+      }
+
+      const path = (usingPagesMap
+        ? // in development we're using a Map, so the value here is a page object
+          (pathOrPageObject as IGatsbyPage).path
+        : // in builds we're using a Set so the page path is the value
+          pathOrPageObject) as string
+
+      const fullPage: IGatsbyPage | undefined = pages.get(path)
+
+      foundOwnerNodeId = fullPage?.ownerNodeId === nodeId
+
+      const foundPageIdInContext = fullPage?.context.id === nodeId
+
+      if (foundOwnerNodeId) {
+        foundPageBy = `ownerNodeId`
+      } else if (foundPageIdInContext && fullPage) {
+        const pageCreatedByPluginName = nodes.get(fullPage.pluginCreatorId)
+          ?.name
+
+        const pageCreatedByFilesystemPlugin =
+          pageCreatedByPluginName === `gatsby-plugin-page-creator`
+
+        foundPageBy = pageCreatedByFilesystemPlugin
+          ? `filesystem-route-api`
+          : `context.id`
+      }
+
+      if (
+        fullPage &&
+        // first check for the ownerNodeId on the page. this is
+        // the defacto owner. Can't get more specific than this
+        (foundOwnerNodeId ||
+          // if there's no specified owner look to see if
+          // pageContext has an `id` variable which matches our
+          // nodeId. Using an "id" as a variable in queries is common
+          // and if we don't have an owner this is a better guess
+          // of an owner than grabbing the first page query we find
+          // that's mapped to this node id.
+          // this also makes this work with the filesystem Route API without
+          // changing that API.
+          foundPageIdInContext)
+      ) {
+        // save this path to use in our manifest!
+        ownerPagePath = fullPage.path
+      }
+    }
+
+    if (ownerPagePath) {
+      pagePath = ownerPagePath
+    }
+  }
+
+  return {
+    page: {
+      path: pagePath || null,
+    },
+    foundPageBy,
+  }
+}
+
+// these id's correspond to error id's in
+// packages/gatsby-cli/src/structured-errors/error-map.ts
+export const foundPageByToLogIds = {
+  none: `11801`,
+  [`context.id`]: `11802`,
```

Comment:
Any reason why you're not doing `contextId` to make the syntax easier to use?
```suggestion
  contextId: `11802`,
```

Reply:
To me it seemed clearer what it is, an id in context rather than something more official, and the syntax here isn't particularly hard to write/read :) I can change it though too

Reply:
I was just curious, you can keep it :)

Reply:
👌 sounds good :)